File: templates/Uwp/Testing/WinAppDriver._VB/Param_ProjectName.Tests.WinAppDriver/BasicTests.vb
Imports System.IO
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports OpenQA.Selenium
Imports OpenQA.Selenium.Appium
Imports OpenQA.Selenium.Appium.Windows
Imports OpenQA.Selenium.Remote

<TestClass>
Public Class BasicTests

    ' TODO WTS: install WinAppDriver and start it before running tests: https://github.com/Microsoft/WinAppDriver
    Protected Const WindowsApplicationDriverUrl As String = "http://127.0.0.1:4723"

    ' TODO WTS: set the app launch ID.
    ' The part before "!App" will be in Package.Appxmanifest > Packaging > Package Family Name.
    ' The app must also be installed (or launched for debugging) for WinAppDriver to be able to launch it.
    Protected Const AppToLaunch As String = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX_XXXXXXXXXXXXX!App"

    Protected Shared Property AppSession As WindowsDriver(Of WindowsElement)

    Private Shared _screenshotFolder As String

    <ClassInitialize>
    Public Shared Sub Setup(context As TestContext)
        ' TODO WTS: change the location where screenshots are saved.
        ' Create separate folders for saving the results of each test run.
        _screenshotFolder = $"{Path.GetPathRoot(Environment.CurrentDirectory)}\\Temp\\Screenshots\\{DateTime.Now.ToString("dd_HHmm")}\\"

        ' Make sure the folder exists or saving screenshots will fail
        If Not Directory.Exists(_screenshotFolder) Then
            Directory.CreateDirectory(_screenshotFolder)
        End If
    End Sub

    <TestInitialize>
    Public Sub LaunchApp()
        If AppSession Is Nothing Then
            Dim appiumOptions = New AppiumOptions()
            appiumOptions.AddAdditionalCapability("app", AppToLaunch)
            appiumOptions.AddAdditionalCapability("deviceName", "WindowsPC")
            AppSession = New WindowsDriver(Of WindowsElement)(New Uri(WindowsApplicationDriverUrl), appiumOptions)

            Assert.IsNotNull(AppSession, "Unable to launch app.")

            AppSession.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(4)

            ' Maximize the window to have a consistent size and position.
            AppSession.Manage().Window.Maximize()
        End If
    End Sub

    ' TODO WTS: Add other tests as appropriate.
    <TestMethod>
    Public Sub TakeScreenshotOfLaunchPage()
        Dim screenshotFileName = Path.Combine(_screenshotFolder, $"{Path.GetRandomFileName()}.png")
        Dim screenshot = AppSession.GetScreenshot()
        screenshot.SaveAsFile(screenshotFileName, ScreenshotImageFormat.Png)
        Assert.IsTrue(File.Exists(screenshotFileName))
    End Sub

    <TestCleanup>
    Public Sub TearDown()
        If AppSession IsNot Nothing Then
            AppSession.Dispose()
            AppSession = Nothing
        End If
    End Sub
End Class

File: code/TemplateStudioForUWP/Templates/Test/WinAppD/Param_ProjectName.Tests.WinAppDriver/BasicTests.cs
using System;
using System.IO;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.Appium;
using OpenQA.Selenium.Appium.Windows;

namespace Param_RootNamespace.Tests.WinAppDriver
{
    [TestClass]
    public class BasicTests
    {
        // TODO WTS: install WinAppDriver and start it before running tests: https://github.com/Microsoft/WinAppDriver
        protected const string WindowsApplicationDriverUrl = "http://127.0.0.1:4723";

        // TODO WTS: set the app launch ID.
        // The part before "!App" will be in Package.Appxmanifest > Packaging > Package Family Name.
        // The app must also be installed (or launched for debugging) for WinAppDriver to be able to launch it.
        protected const string AppToLaunch = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX_XXXXXXXXXXXXX!App";

        protected static WindowsDriver<WindowsElement> AppSession { get; set; }

        private static string _screenshotFolder;

        [ClassInitialize]
        public static void Setup(TestContext context)
        {
            // TODO WTS: change the location where screenshots are saved.
            // Create separate folders for saving the results of each test run.
            _screenshotFolder = $"{Path.GetPathRoot(Environment.CurrentDirectory)}\\Temp\\Screenshots\\{DateTime.Now.ToString("dd_HHmm")}\\";

            // Make sure the folder exists or saving screenshots will fail.
            if (!Directory.Exists(_screenshotFolder))
            {
                Directory.CreateDirectory(_screenshotFolder);
            }
        }

        [TestInitialize]
        public void LaunchApp()
        {
            if (AppSession == null)
            {
                var appiumOptions = new AppiumOptions();
                appiumOptions.AddAdditionalCapability("app", AppToLaunch);
                appiumOptions.AddAdditionalCapability("deviceName", "WindowsPC");
                AppSession = new WindowsDriver<WindowsElement>(new Uri(WindowsApplicationDriverUrl), appiumOptions);

                Assert.IsNotNull(AppSession, "Unable to launch app.");

                AppSession.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(4);

                // Maximize the window to have a consistent size and position.
                AppSession.Manage().Window.Maximize();
            }
        }

        // TODO WTS: Add other tests as appropriate.
        [TestMethod]
        public void TakeScreenshotOfLaunchPage()
        {
            var screenshotFileName = Path.Combine(_screenshotFolder, $"{Path.GetRandomFileName()}.png");

            var screenshot = AppSession.GetScreenshot();
            screenshot.SaveAsFile(screenshotFileName, ScreenshotImageFormat.Png);

            Assert.IsTrue(File.Exists(screenshotFileName));
        }

        [TestCleanup]
        public void TearDown()
        {
            if (AppSession != null)
            {
                AppSession.Dispose();
                AppSession = null;
            }
        }
    }
}
